vim-patch:8.2.3520: cannot define a function for thesaurus completion

mcepl opened the pull request:
Problem:    Cannot define a function for thesaurus completion.
Solution:   Add 'thesaurusfunc'. (Yegappan Lakshmanan, closes vim/vim#8987,
            closes 8950)
https://github.com/vim/vim/commit/160e994d768d03a3c826b58115cde94df8fce607

mcepl committed:
vim-patch:8.2.3520: cannot define a function for thesaurus completion

Problem:    Cannot define a function for thesaurus completion.
Solution:   Add 'thesaurusfunc'. (Yegappan Lakshmanan, closes vim/vim#8987,
            closes 8950)
https://github.com/vim/vim/commit/160e994d768d03a3c826b58115cde94df8fce607

zeertzjq reviewed (commented):
~You forgot to add `BV_THSFU` to `option_defs.h`.~

seandewar reviewed (changes requested):
- There's a few lint failures that `make lint` shouid've caught. https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#coding
- This doesn't compile as per @zeertzjq's comment.

zeertzjq commented:
I think you also need to port patch 8.2.3525.

mcepl committed:
vim-patch:8.2.3521: options completion test fails

Problem:    Options completion test fails.
Solution:   Add 'thesaurusfunc' to the results.
https://github.com/vim/vim/commit/abdcfd1c837e244065d4fe04c7a78abae5af3f7e

mcepl committed:
vim-patch:8.2.3525: option variable name does not match option name

Problem:    Option variable name does not match option name. (Christ van
            Willigen)
Solution:   Rename the variable.
https://github.com/vim/vim/commit/d4c4bfa0078a959ff90ef30288fd31d9d38f23d7

mcepl committed:
vim-patch:8.2.3528: 'thesaurus' and 'thesaurusfunc' do not have the same scope

Problem:    'thesaurus' and 'thesaurusfunc' do not have the same scope.
Solution:   Make 'thesaurusfunc' global-local.
https://github.com/vim/vim/commit/f4d8b76d304dabc39c06d2344cd4c7b28484811b

mcepl commented:
all 8.2.3521, 8.2.3525, and 8.2.3528 should be included.

seandewar commented:
This isn't compiling and lint still needs addressing

seandewar reviewed (commented):
Style comments to help appease the linter

seandewar reviewed (commented):
Many of these comments are probably nit-picks, but functionally I think this PR looks good.
We'll have to see what janlazo thinks :+1:

janlazo commented on src/nvim/edit.c:
```diff
@@ -4096,7 +4111,14 @@ int ins_compl_add_tv(typval_T *const tv, const Direction dir, bool fast)
+}
+
+/// Returns TRUE when using a user-defined function for thesaurus completion.
+static int thesaurus_func_complete(int type)
```
```
/// Returns true when using a user-defined function for thesaurus completion.
static bool thesaurus_func_complete(int type)
```

janlazo replied:
`type` should be an enum/typedef.

mcepl replied:
@janlazo Changing `int` to `bool` is simple, but should I create that `enum` or will we wait on the upstream?

seandewar replied:
Create it yourself. You can do this in a future PR.

mcepl replied:
@seandewar BTW, I am not the author of the original vim patch. But I guess it means, that this conversation can be marked as resolved, right?

janlazo reviewed (approved):
LGTM, ignoring style changes.
Fix the code style, address  `thesaurus_func_complete()`, and I'll merge.

zeertzjq commented on src/nvim/edit.c:
```diff
@@ -4114,11 +4115,10 @@ int ins_compl_add_tv(typval_T *const tv, const Direction dir, bool fast)
 }
 
 /// Returns TRUE when using a user-defined function for thesaurus completion.
```
`TRUE` -> `true`

mcepl replied:
When I do `src/clint.py src/nvim/edit.c` I get 294 errors, should I fix all of those? I don't think it has any relation to this vim patch.

Yes, this `TRUE` is changed.

zeertzjq replied:
> should I fix all of those?

I think you just need to pass lint CI.

mcepl replied:
That I have been passing for some time already.

seandewar commented on src/nvim/edit.c:
```diff
@@ -2899,7 +2900,7 @@ static void ins_compl_dictionaries(char_u *dict_start, char_u *pat, int flags, i
   char_u *buf;
   regmatch_T regmatch;
-  char_u **files;
+  char_u * *files;
```
```suggestion
  char_u **files;
```

These still need changing.

mcepl replied:
Done.

seandewar replied:
The others also need fixing.

seandewar commented on src/nvim/edit.c:
```diff
@@ -2544,7 +2545,7 @@ static void ins_compl_longest_match(compl_T *match)
  * Frees matches[].
  */
-static void ins_compl_add_matches(int num_matches, char_u **matches, int icase)
+static void ins_compl_add_matches(int num_matches, char_u * *matches, int icase)
```
```suggestion
static void ins_compl_add_matches(int num_matches, char_u **matches, int icase)
```

seandewar commented on src/nvim/edit.c:
```diff
@@ -2990,7 +2991,7 @@ static void ins_compl_dictionaries(char_u *dict_start, char_u *pat, int flags, i
 }
 
-static void ins_compl_files(int count, char_u **files, int thesaurus, int flags,
+static void ins_compl_files(int count, char_u * *files, int thesaurus, int flags,
```
```suggestion
static void ins_compl_files(int count, char_u **files, int thesaurus, int flags,
```

seandewar commented on src/nvim/edit.c:
```diff
@@ -4096,7 +4111,14 @@ int ins_compl_add_tv(typval_T *const tv, const Direction dir, bool fast)
   }
   return ins_compl_add((char_u *)word, -1, NULL,
-                       (char_u **)cptext, true, &user_data, dir, flags, dup);
+                       (char_u * *)cptext, true, &user_data, dir, flags, dup);
```
```suggestion
                       (char_u **)cptext, true, &user_data, dir, flags, dup);
```

seandewar commented on src/nvim/edit.c:
```diff
@@ -4116,7 +4138,7 @@ static int ins_compl_get_exp(pos_T *ini)
 
   pos_T *pos;
-  char_u **matches;
+  char_u * *matches;
```
```suggestion
  char_u **matches;
```

mcepl replied:
That's weird. All these four changes seem to the best of my knowledge applied to the current PR.

mcepl commented:
Thank you very much, @janlazo.

janlazo commented:
@mcepl  Thanks for your contribution. 
@seandewar @zeertzjq  Thanks for your continued assistance and your attention to detail. You both made it a lot easier to review this MR.

David-Else commented:
How can I use this in my `init.lua`? I used to have in `init.vim`:
```
func Thesaur(findstart, base)
    if a:findstart
	let line = getline('.')
	let start = col('.') - 1
	while start > 0 && line[start - 1] =~ '\a'
	   let start -= 1
	endwhile
	return start
    else
	let res = []
	let h = ''
	for l in split(system('aiksaurus '.shellescape(a:base)), '\n')
	    if l[:3] == '=== '
	    	let h = substitute(l[4:], ' =*$', '', '')
	    elseif l[0] =~ '\a'
		call extend(res, map(split(l, ', '), {_, val -> {'word': val, 'menu': '('.h.')'}}))
	    endif
	endfor
	return res
    endif
endfunc

set thesaurusfunc=Thesaur   " use the 'thesaurusfunc' for external thesaurus
```
But I can't seem to work out the synax for `vim.opt.thesaurusfunc` in Neovim 0.7. Cheers!

mcepl commented:
I have this:
```vim
func Thesaur(findstart, base)
    if a:findstart
        return searchpos('\<', 'bnW', line('.'))[1] - 1
    endif
    let res = []
    let h = ''
    for l in systemlist('aiksaurus '.shellescape(a:base))
        if l[:3] == '=== '
        let h = '('.substitute(l[4:], ' =*$', ')', '')
        elseif l ==# 'Alphabetically similar known words are: '
        let h = "\U0001f52e"
        elseif l[0] =~ '\a' || (h ==# "\U0001f52e" && l[0] ==# "\t")
        call extend(res, map(split(substitute(l, '^\t', '', ''), ', '),
        \                    {_, val -> {'word': val, 'menu': h}}))
        endif
    endfor
    return res
endfunc

set thesaurusfunc=Thesaur
```

David-Else commented:
@mcepl Can you re-create that in an `init.lua` file, how can I define a vimscript user function and reference it?

mcepl commented:
I probably could, but it is just a waste of time. Put this into `~/.config/nvim/plugin/mythesaurus.vim` and be done with it. If you can persuade me with reasonable arguments this has any disadvantages against your Lua solution, I may try to rewrite it.

David-Else commented:
@mcepl I did the following and it worked, thanks for answering. Lua solution is optimal, but I can't be bothered myself :)

```
vim.cmd [[
func Thesaur(findstart, base)
    if a:findstart
	let line = getline('.')
	let start = col('.') - 1
	while start > 0 && line[start - 1] =~ '\a'
	   let start -= 1
	endwhile
	return start
    else
	let res = []
	let h = ''
	for l in split(system('aiksaurus '.shellescape(a:base)), '\n')
	    if l[:3] == '=== '
	    	let h = substitute(l[4:], ' =*$', '', '')
	    elseif l[0] =~ '\a'
		call extend(res, map(split(l, ', '), {_, val -> {'word': val, 'menu': '('.h.')'}}))
	    endif
	endfor
	return res
    endif
endfunc
]]

vim.opt.thesaurusfunc = 'Thesaur'
```
